felix/iptables/table.go

```diff
@@ -1346,6 +1293,88 @@ func (t *Table) applyUpdates() error {
+	inputBytes := buf.GetBytesAndReset()
+
+	if log.GetLevel() >= log.DebugLevel {
+		// Only convert (potentially very large slice) to string at debug level.
+		inputStr := string(inputBytes)
+		t.logCxt.WithField("iptablesInput", inputStr).Debug("Writing to iptables")
+	}
+
+	var outputBuf, errBuf bytes.Buffer
+	args := []string{"--noflush", "--verbose"}
+	if features.RestoreSupportsLock {
+		// Versions of iptables-restore that support the xtables lock also make it impossible to disable.  Make
+		// sure that we configure it to retry and configure for a short retry interval (the default is to try to
+		// acquire the lock only once).
+		lockTimeout := t.lockTimeout.Seconds()
+		if lockTimeout <= 0 {
+			// Before iptables-restore added lock support, we were able to disable the lock completely, which
+			// was indicated by a value <=0 (and was our default).  Newer versions of iptables-restore require the
+			// lock so we override the default and set it to 10s.
+			lockTimeout = 10
+		}
+		lockProbeMicros := t.lockProbeInterval.Nanoseconds() / 1000
+		timeoutStr := fmt.Sprintf("%.0f", lockTimeout)
+		intervalStr := fmt.Sprintf("%d", lockProbeMicros)
+		args = append(args,
+			"--wait", timeoutStr, // seconds
+			"--wait-interval", intervalStr, // microseconds
+		)
+		log.WithFields(log.Fields{
+			"timeoutSecs":         timeoutStr,
+			"probeIntervalMicros": intervalStr,
+		}).Debug("Using native iptables-restore xtables lock.")
+	}
+	cmd := t.newCmd(t.iptablesRestoreCmd, args...)
+	cmd.SetStdin(bytes.NewReader(inputBytes))
+	cmd.SetStdout(&outputBuf)
+	cmd.SetStderr(&errBuf)
+	countNumRestoreCalls.Inc()
+	// Note: calicoXtablesLock will be a dummy lock if our xtables lock is disabled (i.e. if iptables-restore
+	// supports the xtables lock itself, or if our implementation is disabled by config.
+	t.calicoXtablesLock.Lock()
+	err := cmd.Run()
+	t.calicoXtablesLock.Unlock()
+	if err != nil {
+		// To log out the input, we must convert to string here since, after we return, the buffer can be re-used
+		// (and the logger may convert to string on a background thread).
+		inputStr := string(inputBytes)
+		t.logCxt.WithFields(log.Fields{
+			"output":      outputBuf.String(),
+			"errorOutput": errBuf.String(),
+			"error":       err,
+			"input":       inputStr,
+		}).Warn("Failed to execute ip(6)tables-restore command")
+		t.inSyncWithDataPlane = false
+		countNumRestoreErrors.Inc()
+		return err
+	}
+
+	return nil
+}
+
+// InsertRulesNow insets the given rules immediately without removing or syncing
+// other rules. This is primarily useful when bootstrapping and we cannot wait
+// until we have the full state.
+func (t *Table) InsertRulesNow(chain string, rules []Rule) error {
+	features := t.featureDetector.GetFeatures()
+
+	buf := new(RestoreInputBuilder)
+	buf.StartTransaction(t.Name)
+	for i, r := range rules {
+		buf.WriteLine(r.RenderInsertAtRuleNumber(chain, i+1, "", features))
```

Comment:
Instead of adding the BOOTSTRAP comment outside, how about making this method set it?  Feels like that'd be a nice encapsulation; so the table code would be responsible for making sure bootstrap rules get cleaned up.

The table code has a nice UT harness that simulates iptables, would be good to have UT of the new function.

I don't think it's too important, but you could go the whole hog and calculate the proper rule hashes here.  Then if they match correctly later, we'll avoid churning the rules.

Reply:
> The table code has a nice UT harness that simulates iptables, would be good to have UT of the new function.

The harness needed some fixing ;-)